Homework/5

Laimonukas opened the pull request:
Added functions:
Add person (POST),
Get all persons (GET),
Get person by ID (GET),
Adding pets to person (PUT),
Remove person (DELETE),
Remove pet (PUT),
Update person (PATCH),
Marry persons (PUT)

Laimonukas committed:
Homework/5

Added functions:
Add person (POST),
Get all persons (GET),
Get person by ID (GET),
Adding pets to person (PUT),
Remove person (DELETE),
Remove pet (PUT),
Update person (PATCH),
Marry persons (PUT)

Almantask commented on Homework5/Controllers/PeopleApiController.cs:
```diff
@@ -0,0 +1,198 @@
+            }
+
+            RemovePetObjectFromPersonObject(tempPersonObj, tempPetObj);
+            tempPersonObj = FindPersonById(id);
```
Why do you need to find a person again?

Laimonukas replied:
I wanted to return an actual changed value, and first time finding the object was to get the reference. This is actually unneeded.

Almantask commented on Homework5/Controllers/PeopleApiController.cs:
```diff
@@ -0,0 +1,198 @@
+        public IActionResult RemovePetFromPerson(int id, string petName)
+        {
+            Person tempPersonObj = FindPersonById(id);
+            if (tempPersonObj == null)
```
temp and Obj in the name are not necessary, they don't add value nor clarify intention. Usually we use those as intermediate values, but in this case they are the actuals.

Almantask commented on Homework5/Controllers/PeopleApiController.cs:
```diff
@@ -0,0 +1,198 @@
+            {
+                return NotFound($"Person by id {id} was not found.");
+            }
+            ChangeSpouseIds(id, spouseId);
+
```
Move this to person class.

Laimonukas replied:
I don't understand why this needs to be moved to Person class, as this function changes two different objects so it still needs to query through available object list to get correct spouses. Also it would require a list of objects be passed to when in person class. Unless I'm missing something obvious

Almantask replied:
You're right. Another person had Marry function or similar, which changed ids of already found person and existing person to update them accordingly. I think you could spli this function into finding the person and then updating the ids on both through that method. Updating two ids is very simple and does not reduce complexity, but method called `Mary` clarifies the intention of two id changes for spouses.

Almantask commented on Homework5/Controllers/PeopleApiController.cs:
```diff
@@ -0,0 +1,198 @@
+            return Ok();
+        }
+
+        /// <summary>
+        /// Returns found Pet object when provided a List of Pet object and search name
+        /// </summary>
+        /// <param name="petList">List to search in</param>
+        /// <param name="petName">Name to search for</param>
+        /// <returns></returns>
```
Instead of focusing on documenting private methods- do this for your public ones.

Almantask commented on Homework5/Controllers/PeopleApiController.cs:
```diff
@@ -0,0 +1,198 @@
+                        personList[i] = newPersonObj;
+                    }
+                }
+            }
```
This does not really update a person, but rather swap the element of a person list. Not quite what was asked.
Also, don't rely on matching references, simply set all properties of the old object to properties of the new object.

Almantask replied:
To find the person you need you will need an equality by id.

Almantask replied:
Could also be moved to a person class, because it can manage itself just fine without relying on anything. Controller should be as thin as possible.

Laimonukas committed:
Update to PersonApi

Changed tempObj names to remove temp.

Moved and changed Update person function.

Almantask commented on Homework5/Controllers/PeopleApiController.cs:
```diff
@@ -39,63 +39,63 @@ public IActionResult GetAllPersons()
 
-            Pet tempPetObj = FindPetByName(tempPersonObj.Pets, petName);
-            if (tempPetObj == null)
+            Pet petObj = FindPetByName(personObj.Pets, petName);
```
`Obj` is noise. It doesn't add to clarity.

Almantask replied:
So it can be removed from a variable name.

Almantask commented on Homework5/Person.cs:
```diff
@@ -12,5 +12,18 @@ public class Person
+
+
+        public void UpdatePerson(Person newDetails) {
+            if (Id != newDetails.Id) Id = newDetails.Id;
+
+            if (SpouseId != newDetails.SpouseId) SpouseId = newDetails.SpouseId;
+
+            if (Name != newDetails.Name) Name = newDetails.Name;
+
+            if (Age != newDetails.Age) Age = newDetails.Age;
+
+            if (Pets != newDetails.Pets) Pets = newDetails.Pets;
```
You can make it more simple- just assign the value. If it is not the same- it will be the same. If it was the same- nothing will change when you set the same value to the same value.

Almantask replied:
So remove the ifs.
